handlers/handlers.go

```diff
@@ -0,0 +1,244 @@
+package handlers
+
+import (
+	"encoding/json"
+	"gin/models"
+	"github.com/gin-gonic/gin"
+	"github.com/olivere/elastic/v7"
+	"html"
+	"net/http"
+	"strconv"
+)
+
+const INDEX_NAME = "books"
+
+func CreateBook(c *gin.Context) {
+	var book models.Book
+	if err := c.ShouldBindJSON(&book); err != nil {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": err.Error()})
+	}
+
+	doc, err := models.ElasticClient.Index().Index(INDEX_NAME).BodyJson(book).Do(c)
+
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, gin.H{
+		"status": "created",
+		"id":     doc.Id,
+	})
+}
+
+func UpdateBookTitleById(c *gin.Context) {
+	id := c.Param("id")
+
+	var book *models.Book
+	if err := c.ShouldBindJSON(&book); err != nil {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": err.Error()})
+		return
+	}
+
+	_, err := models.ElasticClient.
+		Update().
+		Index(INDEX_NAME).
+		Id(id).
+		Doc(gin.H{"title": book.Title}).
+		Do(c)
+
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
+		return
+	}
+	c.JSON(http.StatusOK, gin.H{"status": "updated"})
+}
+
+func GetBookById(c *gin.Context) {
+	id := c.Param("id")
+
+	doc, err := models.ElasticClient.
+		Get().
+		Index(INDEX_NAME).
+		Id(id).
+		Do(c)
+
+	if err != nil {
+		elasticErr := err.(*elastic.Error)
+
+		if elasticErr != nil {
+			c.AbortWithStatusJSON(elasticErr.Status, gin.H{"message": elasticErr.Error()})
+			return
+		}
+
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": err.Error()})
+		return
+	}
+
+	var book models.Book
+	err = json.Unmarshal(doc.Source, &book)
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{"message": err.Error()})
+	}
+
+	c.JSON(http.StatusOK, book)
+}
+
+func DeleteBookById(c *gin.Context) {
+	id := c.Param("id")
+
+	doc, err := models.ElasticClient.
+		Delete().
+		Index(INDEX_NAME).
+		Id(id).
+		Do(c)
+
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": err.Error()})
+		return
+	}
+
+	c.JSON(http.StatusOK, gin.H{"status": "deleted", "id": doc.Id})
+}
+
+func SearchBooks(c *gin.Context) {
+	title := c.Query("title")
+	authorName := c.Query("author_name")
+	minPrice, minPriceOk := c.GetQuery("min_price")
+	maxPrice, maxPriceOk := c.GetQuery("max_price")
+
+	if title == "" && authorName == "" && minPrice == "" && maxPrice == "" {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{"message": "at least one query parameter is required"})
+		return
+	}
+
+	index := models.ElasticClient.Search().Index(INDEX_NAME).Pretty(false).Size(10000)
+
+	boolQuery := elastic.NewBoolQuery()
```

Comment:
`http://localhost:8080/search?title=do`

I am not sure I understand this comment. This query returns all books with a title that contains the word "do". And in general this words for any substring of complete words. It doesn't work for general substring. This could be fixed using a wildcard query of a different analyzer for the title field.

Reply:
This should return only books whose title is "do", not contains do.